---
output: html_document
---
## Simulation of responses to selection with changes in recombination rate and landscape.
#### Susan E. Johnston, `r Sys.time()`

### Introduction.

Meiotic recombination is a fundamental feature of sexual reproduction and is an important source of genetic diversity. It can uncouple beneficial alleles from linked deleterious ones, creating new combinations of alleles that can allow populations to respond faster to selection. On the other hand, recombination can increase mutations at crossover sites, and can break up favourable combinations of alleles previously built up by selection. Recombination rate is likely to have a fairly simple genetic architecture, with a handful of loci with relatively large effects on rate variation, such as the *RNF212*/*CPLX1* region in humans, cattle and sheep.

The majority of cross-overs occur in regions known as “hotspots” – short stretches of DNA 1-2kb long. In some mammal species (e.g. humans, mice) the location of hotspots can change rapidly, often attributed to variation in the gene *PRDM9*; in others (e.g. dogs), *PRDM9* has lost its function leading to stable recombination hotspots over longer periods of time. However, there remains little understanding of how recombination landscapes vary in non-model species, and if rapid hotspot turnover could be favoured if selection is strong (e.g. in domesticated species).

The following is a simulation study that investigates how responses to selection change when recombination landscape remains stable, incomparison to when it changes due to the action of a single polymorphic modifier locus, such as *PRDM9*.

### Model

```{r echo = F, message=FALSE}

library(ggplot2)
library(plyr)
library(data.table)

sapply(paste0("R/", dir("R", pattern = ".R")), source)

```

The study simulates selection on a polygenic trait in a mammal population with variation in recombination landscape. The simulations described in this report are modelled on selection regimes in domestic cattle, where males with highest trait values are selected to sire offspring in subsequent generations. 

##### Founder population
The founder population can be specified in the following example.

```{r}

n.females         <- 50  # Number of females in the founder generation
n.males           <- 50  # Number of males in the founder generation
n.offspring       <- 2   # Number of offspring per female
male.sel.thresh   <- 0.4 # Top proportion of males that will be selected
female.sel.thresh <- 1   # Top proportion of females that will be selected

```

Here, we will have a founder population of `r n.females` females and `r n.males` males. Each female will have `r n.offspring`. Males that have trait values in the top `r male.sel.thresh*100`% of all males are selected for mating. At present, all females are mated. In running the model, we will force each female to have one offspring of each sex by specifying `force.equal.sex <- TRUE`. If this is specified as `FALSE` then the sex of the offspring is sampled at random. We will also `force.equal.male.success <- TRUE` which will sample males at random, but without replacement, with as equal a contribution of males as possible. If specified as `FALSE` then sires will be sampled randomly from all males above the selection threshold.

##### Specification of recombination landscape and modifier loci.

I will start with a simple model of recombination on a chromosome that is 1 Morgan long (i.e. an average of one crossover per meiosis). The number of loci contributing to the trait are specified. Each locus has two alleles: `0` resulting in no change in phenotype, and `1` that results in an increase in phenotype by one unit. 

```{r}
n.loci <- 101   # The number of loci contributing to the trait
```

The function then requires information on the recombination fraction between loci, as well as the allele frequencies of the `0` alleles at each locus. Lets make 

```{r}
map.1 <- 

```



iterations <- 5
no.founder.hap <- 100
generations <- 100





### Future studies and considerations.

There are several assumptions made in the current model that may not hold in a true mammal population. For example:

* At present I assume equal recombination rates in each sex.
* Modifier loci for recombination rate e.g. *RNF212*/*CLPX1*  may have sex specific or sexually antagonistic effects.
* Population sizes are allowed to vary naturally, which can result in extinctions or very large increases in population size. 





#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
# 1. Define simulation parameters and sampling distributions   #
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#


generations <- 100
no.females <- 50
no.males <- 50
no.loci <- 100
no.offspring <- 2
no.founder.hap <- 100
male.sel.thresh <- 0.4
iterations <- 5
sampling.int <- 10


#~~ Create the sampling distrubutions

map.dist <- diff(map$cM.Position.Sex.Averaged[seq(1, nrow(map), sampling.int)])
map.dist <- map.dist[which(map.dist >= 0 & map.dist < 2)]
maf.info <- map$MAF

#~~ Create object for results

res.list.modifier.present <- list()
res.list.modifier.absent  <- list()

system.time({
  
  for(i in 1:iterations){
    
    print(paste("Running simulation", i))
    
    #     xr1   <- sample(map.dist/100, replace = T, no.loci)
    #     xr2   <- sample(map.dist/100, replace = T, no.loci)
    #     xrhet <- (xr1 + xr2)/2
    
    xr1   <- sample(map.dist/100, replace = T, no.loci)
    xr2   <- xr1*2
    xrhet <- (xr1 + xr2)/2
    
    map.list <- list(xr1, xrhet, xr2)
    rm(xr1, xrhet, xr2)
    
    allele.freqs <- map$MAF
    allele.freqs <- sample(allele.freqs, no.loci)
    allele.freqs <- allele.freqs + (runif(no.loci) < 0.5)/2
    
    
    x1 <- createFounderObject(map.list = map.list, allele.freqs = allele.freqs, n.found.hap = no.founder.hap,
                              n.loci = no.loci, n.f = no.females, n.m = no.males, f.RS = no.offspring, f.RS.Pr = NULL)
    
    res.list.modifier.present[[i]] <- rbindlist(
      simPopulationResponse(map.list = map.list, allele.freqs = allele.freqs,
                            n.found.hap = no.founder.hap, n.loci = no.loci, 
                            n.f = no.females, n.m = no.males, f.RS = no.offspring, 
                            sel.thresh.f = 1, sel.thresh.m = male.sel.thresh,
                            modifier.found.freq = 0.4, n.generations = generations,
                            FounderObject = x1)$results)
    
    res.list.modifier.present[[i]]$Simulation <- i
    
    res.list.modifier.absent[[i]] <- rbindlist(
      simPopulationResponse(map.list = map.list, allele.freqs = allele.freqs,
                            n.found.hap = no.founder.hap, n.loci = no.loci, 
                            n.f = no.females, n.m = no.males, f.RS = no.offspring, 
                            sel.thresh.f = 1, sel.thresh.m = male.sel.thresh,
                            modifier.found.freq = 0, n.generations = generations,
                            FounderObject = x1)$results)
    
    res.list.modifier.absent[[i]]$Simulation <- i
  }
})

save(res.list.modifier.present, res.list.modifier.absent, 
     file = paste0("results/g", generations, "_it", iterations,  "_f", no.females, "_m", no.males, "_o", no.offspring,
                   "_l", no.loci, "_msel", male.sel.thresh, ".Rdata"))




# load(paste0("results/g", generations, "_it", iterations,  "_f", no.females, "_m", no.males, "_o", no.offspring,
#             "_l", no.loci, "_msel", male.sel.thresh, ".Rdata"))

test <- rbind(cbind(Modifier.Start = 0.4, rbindlist(res.list.modifier.present)),
              cbind(Modifier.Start = 0  , rbindlist(res.list.modifier.absent)))

test$Simulation <- as.factor(test$Simulation)
test$Modifier.Start <- as.factor(test$Modifier.Start)
head(test)

df1 <- ddply(test, .(GEN, Modifier.Start, Simulation), summarise, MeanPHENO = mean(PHENO))

df3 <- ddply(test, .(GEN, Modifier.Start, Simulation), summarise, VarPHENO = var(PHENO))

setkey(test, GEN, Modifier.Start, Simulation)

testfunc <- function(vec) {sum(vec - 1)/(2*length(vec))}

df2 <- test[,list(MeanPHENO=mean(PHENO),
                  VarPheno = var(PHENO),
                  PopSize = length(PHENO),
                  Modifier.Freq = testfunc(modifier)),
            by=list(GEN, Modifier.Start, Simulation)] 

head(df2)
test


ggplot(df2, aes(GEN, MeanPHENO, col = Modifier.Start, group = interaction(Simulation, Modifier.Start))) + 
  geom_line()

ggplot(df2, aes(GEN, VarPheno, col = Modifier.Start, group = interaction(Simulation, Modifier.Start))) + 
  geom_line()

ggplot(df2, aes(GEN, PopSize, col = Modifier.Start, group = interaction(Simulation, Modifier.Start))) + 
  geom_line()

ggplot(df2, aes(GEN, Modifier.Freq, col = Modifier.Start, group = interaction(Simulation, Modifier.Start))) + 
  geom_line()



ggplot(df1, aes(GEN, MeanPHENO, col = Modifier.Start)) +
  geom_point(alpha = 0) +
  stat_smooth() +
  theme(axis.text.x  = element_text (size = 16, vjust = 1),
        axis.text.y  = element_text (size = 16, hjust = 1),
        strip.text.x = element_text (size = 16, vjust = 0.7),
        axis.title.y = element_text (size = 16, angle = 90, vjust = 1),
        axis.title.x = element_text (size = 16, vjust = 0.2),
        strip.background = element_blank()) +
  scale_colour_brewer(palette = "Set1") +
  labs(title = paste0("g", generations, "_it", iterations,  "_f", no.females, "_m", no.males, "_o", no.offspring,
                      "_l", no.loci, "_msel", male.sel.thresh))

beepr::beep()